C++ and Rust together killed the competition

acking-you opened the pull request:
## What changes were made

- Added performance test code for FFT in C++. The performance issue in C++ was resolved by introducing mimalloc through vcpkg, and the performance has reached the top level together with Rust, thanks to the guidance of @YuHuanTin. Close #9 

- The Rust code simply used mimalloc as the memory allocator based on #11, and at this point, it completely dominated the competition, outperforming the primary language promoted in this repository by over 50%.


## Benchmark results
8-core AMD 7700, Ubuntu 24.04
```
==========================================================================================
Benchmark Summary (execution time in ms)
==========================================================================================
Program                Input   Runs      Fastest      Slowest       Median      Average
---------------------------------------------------------------------------------------
rust                      18     10    10.476 ms    15.138 ms    11.079 ms    11.588 ms
moonbit                   18     10    17.850 ms    22.983 ms    18.758 ms    19.264 ms
go                        18     10    28.110 ms    32.182 ms    30.194 ms    30.019 ms
cpp                       18     10    11.366 ms    12.758 ms    11.843 ms    11.920 ms
---------------------------------------------------------------------------------------
rust                      20     10    46.827 ms    57.326 ms    48.651 ms    49.293 ms
moonbit                   20     10    79.788 ms    85.990 ms    81.207 ms    82.172 ms
go                        20     10   118.245 ms   136.685 ms   121.474 ms   124.441 ms
cpp                       20     10    50.378 ms    61.190 ms    51.744 ms    52.566 ms
---------------------------------------------------------------------------------------
rust                      22     10   209.130 ms   261.543 ms   214.435 ms   217.823 ms
moonbit                   22     10   349.026 ms   399.255 ms   357.258 ms   364.601 ms
go                        22     10   506.437 ms   610.093 ms   513.382 ms   521.883 ms
cpp                       22     10   234.986 ms   267.293 ms   241.498 ms   243.606 ms
==========================================================================================
```


<img width="1280" height="960" alt="bench_avg" src="https://github.com/user-attachments/assets/3241d0a9-7d12-40f9-a154-ecb1b092cdff" />

acking-you committed:
Improve rust code

acking-you committed:
add cpp support && Impl The most efficient Rust code

YuHuanTin commented:
C++ Acceleration Methods
1. Use libc++ (clang)
2. Add the `-ffast-math` parameter (accelerates `sin`, `cos` calculations)
3. Use `mimalloc` to accelerate memory allocation (static linking results in faster runtime)
4. Use

```c++
        std::vector<Complex> a0, a1;
        a0.reserve(n / 2);
        a1.reserve(n / 2);

        for (int i = 0; i < n; i += 2)
        {
            a0.emplace_back(arr[i]);
        }

        for (int i = 1; i < n; i += 2)
        {
            a1.emplace_back(arr[i]);
        }
```
Instead of
```c++
        std::vector<Complex> a0, a1;
        a0.reserve(n / 2);
        a1.reserve(n / 2);
        
        // Use ranges with stride to get even/odd elements
        auto even_indices = std::views::iota(0uz, n) | std::views::filter([](size_t i) { return i % 2 == 0; });
        auto odd_indices = std::views::iota(0uz, n) | std::views::filter([](size_t i) { return i % 2 == 1; });
        
        std::ranges::transform(even_indices, std::back_inserter(a0), [&arr](size_t i) { return arr[i]; });
        std::ranges::transform(odd_indices, std::back_inserter(a1), [&arr](size_t i) { return arr[i]; });
```
--

Using Clang++20, LLVM-20
```CMakeLists.txt
cmake_minimum_required(VERSION 3.20)
project(fft_benchmark)

# Set C++23 standard
set(CMAKE_CXX_STANDARD 23)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS OFF)

add_compile_options(-stdlib=libc++ -static)
add_link_options(-fuse-ld=lld-20 -static)

# Enable native CPU optimizations
set(CMAKE_CXX_FLAGS_RELEASE "-O3 -ffast-math -march=native -mtune=native -DNDEBUG")

# Add executable
add_executable(main src/main.cpp)

target_link_libraries(main PRIVATE c++ c++abi /path/to/release/static-build/libmimalloc.a)

# Set output directory
set_target_properties(main PROPERTIES
    RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin
)
```

dynamic link
<img width="1182" height="885" alt="image" src="https://github.com/user-attachments/assets/5bf53268-0815-4b39-81a9-a0e654346fa7" />
```
Benchmark Summary (execution time in ms)
==========================================================================================
Program                Input   Runs      Fastest      Slowest       Median      Average
---------------------------------------------------------------------------------------
rust                      18     10    10.447 ms    10.711 ms    10.512 ms    10.534 ms
moonbit                   18     10    16.383 ms    17.616 ms    16.817 ms    16.877 ms
cpp                       18     10    10.966 ms    11.327 ms    11.092 ms    11.109 ms
---------------------------------------------------------------------------------------
rust                      20     10    45.896 ms    46.722 ms    46.397 ms    46.355 ms
moonbit                   20     10    74.209 ms    77.384 ms    75.263 ms    75.506 ms
cpp                       20     10    47.061 ms    49.241 ms    47.539 ms    47.751 ms
---------------------------------------------------------------------------------------
rust                      22     10   205.100 ms   208.441 ms   206.621 ms   206.621 ms
moonbit                   22     10   325.902 ms   343.097 ms   328.962 ms   330.863 ms
cpp                       22     10   211.397 ms   218.943 ms   216.021 ms   215.455 ms
==========================================================================================
```

static link
<img width="1184" height="890" alt="image" src="https://github.com/user-attachments/assets/eea8c191-dad9-4e06-bdad-68388ae2d75b" />
```
==========================================================================================
Benchmark Summary (execution time in ms)
==========================================================================================
Program                Input   Runs      Fastest      Slowest       Median      Average
---------------------------------------------------------------------------------------
rust                      18     10    10.398 ms    10.792 ms    10.497 ms    10.526 ms
moonbit                   18     10    16.583 ms    17.489 ms    16.881 ms    16.989 ms
cpp                       18     10    10.294 ms    10.890 ms    10.365 ms    10.451 ms
---------------------------------------------------------------------------------------
rust                      20     10    46.046 ms    47.070 ms    46.392 ms    46.410 ms
moonbit                   20     10    74.367 ms    82.793 ms    75.304 ms    76.168 ms
cpp                       20     10    45.288 ms    46.902 ms    45.988 ms    46.065 ms
---------------------------------------------------------------------------------------
rust                      22     10   205.143 ms   214.420 ms   208.115 ms   208.413 ms
moonbit                   22     10   324.497 ms   339.519 ms   328.718 ms   330.566 ms
cpp                       22     10   202.836 ms   207.198 ms   205.516 ms   205.437 ms
==========================================================================================
```

acking-you commented:
> C++ Acceleration Methods
> 
> 1. Use libc++ (clang)
> 2. Add the `-ffast-math` parameter (accelerates `sin`, `cos` calculations)
> 3. Use `mimalloc` to accelerate memory allocation (static linking results in faster runtime)
> 4. Use
> 
> ```c++
>         std::vector<Complex> a0, a1;
>         a0.reserve(n / 2);
>         a1.reserve(n / 2);
> 
>         for (int i = 0; i < n; i += 2)
>         {
>             a0.emplace_back(arr[i]);
>         }
> 
>         for (int i = 1; i < n; i += 2)
>         {
>             a1.emplace_back(arr[i]);
>         }
> ```
> 
> Instead of
> 
> ```c++
>         std::vector<Complex> a0, a1;
>         a0.reserve(n / 2);
>         a1.reserve(n / 2);
>         
>         // Use ranges with stride to get even/odd elements
>         auto even_indices = std::views::iota(0uz, n) | std::views::filter([](size_t i) { return i % 2 == 0; });
>         auto odd_indices = std::views::iota(0uz, n) | std::views::filter([](size_t i) { return i % 2 == 1; });
>         
>         std::ranges::transform(even_indices, std::back_inserter(a0), [&arr](size_t i) { return arr[i]; });
>         std::ranges::transform(odd_indices, std::back_inserter(a1), [&arr](size_t i) { return arr[i]; });
> ```


Thank you very much, this is a very interesting optimization!

acking-you commented:
@YuHuanTin  I'll add all these during my spare time to this PR. Thanks again.

acking-you committed:
optimize cpp code with mimalloc

acking-you commented:
@YuHuanTin Hello, I have already run the latest results locally based on your optimizations and changes, but C++ seems unable to reproduce the performance in your environment. The specific results are in: https://github.com/moonbit-community/benchmark-fft/pull/12#issue-3407600052

I made the following changes to the C++ code:  
1. Introduced static linking of mimalloc based on vcpkg's manifest mode.  
2. Modified the C++ code style and compilation parameters, and enforced the use of the clang-20 compiler.  
3. I even forcibly added the following hard-coded configurations to ensure static linking with the correct libc++, but it barely improved performance, so I removed this option.
```
add_compile_options(-stdlib=libc++ -static)
add_link_options(-fuse-ld=lld-20 -static)

# Use mimalloc and force to use llvm-20 libc++
target_link_libraries(main PRIVATE 
    $<IF:$<TARGET_EXISTS:mimalloc-static>,mimalloc-static,mimalloc>
    /usr/lib/llvm-20/lib/libc++.a
    /usr/lib/llvm-20/lib/libc++abi.a
)
```

But even so, I still couldn't reproduce the benchmark results you achieved, which is quite strange.

To run my current C++ code locally on your machine, follow these steps:  
1. Run `setup-cpp-deps.sh` to set up the vcpkg environment in one step.  
2. Execute `cd fft/cpp && cmake --preset=vcpkg && cmake --build --preset=vcpkg` (make sure to source beforehand to ensure $VCPKG_ROOT is set).

YuHuanTin commented:
vcpkg.json should be:
```
{
  "dependencies": [
    {
      "name": "mimalloc",
      "features": [
        "override"
      ]
    }
  ]
}
```
`mimalloc` requires the `override` feature to override the implementation of `new` and `delete`.
https://vcpkg.link/ports/mimalloc#features

Next, you'll get my first image. If you want to speed things up to the second image, you'll need to use `libc++` and statically link it ( if you don't statically link, it will be slower).
At the same time, static linking doesn't seem to speed up `libstdc++`
(at least on my end, libc++ is a bit faster than libstdc++).

acking-you commented:
> `mimalloc` requires the `override` feature to override the implementation of `new` and `delete`. https://vcpkg.link/ports/mimalloc#features

Thanks for the reminder. Now that the tests for C++ and Rust are on the same level, I should change the title. 🤣 "C++ and Rust together killed the competition".

acking-you committed:
fix cpp mimalloc

acking-you committed:
back to edition 2024

acking-you committed:
add more flags to optimize performance
